src/ngraph/pass/nop_elimination.cpp

```diff
@@ -33,76 +38,156 @@
 using namespace std;
 using namespace ngraph;
 
-#define TI(x) std::type_index(typeid(x))
+#define TI(x) x::type_info
+
+static bool remove_update_name(const std::shared_ptr<Node>& node,
+                               const std::shared_ptr<Node>& node_input)
+{
+    bool has_result_output = false;
+    for (auto& output : node->output(0).get_target_inputs())
+    {
+        if (as_type<op::Result>(output.get_node()))
+        {
+            has_result_output = true;
+        }
+    }
+    // ignore trivial elimination
+    if (has_result_output && as_type_ptr<ngraph::op::Parameter>(node_input))
+    {
+        return false;
+    }
+    if (!has_result_output || node_input->get_users().size() == 1)
+    {
+        node_input->set_friendly_name(node->get_friendly_name());
+        replace_node(node, node_input);
+        return true;
+    }
+    return false;
+}
 
 static bool eliminate_pad(const std::shared_ptr<Node>& node)
 {
-    auto pad = std::static_pointer_cast<op::Pad>(node);
+    auto pad = std::static_pointer_cast<op::v0::Pad>(node);
+    // skip if shapes are dynamic
+    if (pad->get_input_partial_shape(0).is_dynamic() ||
+        pad->get_output_partial_shape(0).is_dynamic())
+    {
+        return false;
+    }
+
     if (pad->get_input_shape(0) == pad->get_output_shape(0))
     {
-        replace_node(node, node->get_argument(0));
-        return true;
+        return remove_update_name(node, node->get_argument(0));
     }
     return false;
 }
 
 static bool eliminate_sum(const std::shared_ptr<Node>& node)
 {
-    auto sum = std::static_pointer_cast<op::Sum>(node);
+    auto sum = std::static_pointer_cast<op::v0::Sum>(node);
     if (sum->get_reduction_axes().empty())
     {
-        replace_node(node, node->get_argument(0));
-        return true;
+        return remove_update_name(node, node->get_argument(0));
     }
     return false;
 }
 
 static bool eliminate_convert(const std::shared_ptr<Node>& node)
 {
-    auto convert = std::static_pointer_cast<op::Convert>(node);
-    if (convert->get_convert_element_type() == convert->get_argument(0)->get_element_type())
+    bool is_out_type_agnostic = false;
+    static const std::set<NodeTypeInfo> type_agnostic{TI(op::v3::NonZero)};
+    if (node->output(0).get_target_inputs().size() == 1)
```

Comment:
we could do this for any number of inputs, if they are all the same op? I saw an use case of gather op.

Reply:
yes ,the static lookup table `type_agnostic` allows us to add more ops (currently just have NonZero, but we could add others, whose compute-logic/result is independent of type of input)

Reply:
what I meant is, even if we have more than 1 consumers of the convert, if they are all `type_agnostic`, we can eliminate the convert

Reply:
this PR tries to handle limited cases, along with other op eliminations. may be we could do this as another PR, also could you add a unit-test so in new PR we can test this use-case.

Reply:
yes!

Reply:
this use case will be a different PR